crates/relayer-types/src/core/ics04_channel/timeout.rs

```diff
@@ -204,31 +206,73 @@ pub enum UpgradeTimeout {
     Both(Height, Timestamp),
 }
 
-impl UpgradeTimeout {
+impl Display for Timeout {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
+        match self {
+            Self::Height(height) => write!(f, "{height}"),
+            Self::Timestamp(timestamp) => write!(f, "{timestamp}"),
+            Self::Both(height, timestamp) => write!(f, "{height}, {timestamp}"),
+        }
+    }
+}
+
+impl Timeout {
     pub fn new(height: Option<Height>, timestamp: Option<Timestamp>) -> Result<Self, ChannelError> {
         match (height, timestamp) {
-            (Some(height), None) => Ok(UpgradeTimeout::Height(height)),
-            (None, Some(timestamp)) => Ok(UpgradeTimeout::Timestamp(timestamp)),
-            (Some(height), Some(timestamp)) => Ok(UpgradeTimeout::Both(height, timestamp)),
+            (Some(height), None) => Ok(Timeout::Height(height)),
+            (None, Some(timestamp)) => Ok(Timeout::Timestamp(timestamp)),
+            (Some(height), Some(timestamp)) => Ok(Timeout::Both(height, timestamp)),
             (None, None) => Err(ChannelError::missing_upgrade_timeout()),
         }
     }
 
     pub fn into_tuple(self) -> (Option<Height>, Option<Timestamp>) {
         match self {
-            UpgradeTimeout::Height(height) => (Some(height), None),
-            UpgradeTimeout::Timestamp(timestamp) => (None, Some(timestamp)),
-            UpgradeTimeout::Both(height, timestamp) => (Some(height), Some(timestamp)),
+            Timeout::Height(height) => (Some(height), None),
+            Timeout::Timestamp(timestamp) => (None, Some(timestamp)),
+            Timeout::Both(height, timestamp) => (Some(height), Some(timestamp)),
+        }
+    }
+}
+
+define_error! {
+    #[derive(Debug, PartialEq, Eq)]
+    TimeoutError {
+        InvalidTimestamp
+            { timestamp: String }
+            [ TraceError<ParseTimestampError> ]
+            |e| { format_args!("cannot convert into a `Timestamp` type from string {0}", e.timestamp) },
+
+        InvalidTimeout
+            { timeout: String }
+            |e| { format_args!("invalid timeout {0}", e.timeout) },
+    }
+}
+
+impl FromStr for Timeout {
+    type Err = TimeoutError;
+
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        let split: Vec<&str> = value.split(' ').collect();
+
+        if split.len() != 2 {
+            return Err(TimeoutError::invalid_timeout(value.to_owned()));
         }
+
+        // only timeout timestamp are supported at the moment
+        return split[1]
+            .parse::<Timestamp>()
+            .map(|timestamp| Timeout::Timestamp(timestamp))
+            .map_err(|e| TimeoutError::invalid_timestamp(value.to_owned(), e));
     }
```

Comment:
I haven't tested this parsing logic yet... If anyone has any recommendations or examples of how to do it, please let me know!

Reply:
Do you have examples of the timeout format? If so then we can add a module at the end of the file with a unit test, eg.

```rust
#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn parse_timeout_timestamp() {
     let tests = [
       ("timestamp1", Timeout::new(..., ...)),
       ("timestamp2", Timeout::new(..., ...)),
       // etc.
     ];

     for (timestamp, expected) in tests {
       let parsed = Timeout::from_str(timestamp).unwrap();
       assert_eq!(parsed, expected);
     }
  }
}
```

Reply:
Thanks, @romac! I will add something like this!